File: leetcode/3100-3200/3193. 统计逆序对的数目.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from operator import *
from typing import *
from queue import *
from string import *
from random import *
from sortedcontainers import *

DIRS = [(-1, 0), (1, 0), (0, -1), (0, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float("inf")
MOD = int(1e9 + 7)


class Solution:
    def numberOfPermutations(self, n: int, requirements: List[List[int]]) -> int:
        req = {end: inv for end, inv in requirements}
        if 0 in req and req[0]:
            return 0
        max_k = 400

        dp = [0] * (max_k + 1)
        dp[0] = 1

        for i in range(2, n + 1):
            next_dp = [0] * (max_k + 1)
            sm = 0

            for j in range(max_k + 1):
                sm = (sm + dp[j]) % MOD

                if j >= i:
                    sm = (sm - dp[j - i] + MOD) % MOD

                next_dp[j] = sm

            if (i - 1) in req:
                req_k = req[i - 1]
                for j in range(max_k + 1):
                    if j != req_k:
                        next_dp[j] = 0

            dp = next_dp

        return dp[req.get(n - 1)]

File: leetcode/3100-3200/3193. 统计逆序对的数目.cpp
class Solution {
    const int MOD = 1'000'000'007;

public:
    int numberOfPermutations(int n, vector<vector<int>> &requirements) {
        vector<int> req(n, -1);
        req[0] = 0;
        for (auto &p: requirements) {
            req[p[0]] = p[1];
        }
        if (req[0]) {
            return 0;
        }

        int m = ranges::max(req);
        vector<int> f(m + 1);
        f[0] = 1;
        for (int i = 1; i < n; i++) {
            int mx = req[i] < 0 ? m : req[i];
            if (int r = req[i - 1]; r >= 0) {
                fill(f.begin(), f.begin() + r, 0);
                fill(f.begin() + r + 1, f.begin() + min(i + r, mx) + 1, f[r]);
                fill(f.begin() + min(i + r, mx) + 1, f.end(), 0);
            } else {
                for (int j = 1; j <= mx; j++) { // 计算前缀和
                    f[j] = (f[j] + f[j - 1]) % MOD;
                }
                for (int j = mx; j > i; j--) { // 计算子数组和
                    f[j] = (f[j] - f[j - i - 1] + MOD) % MOD;
                }
            }
        }
        return f[req[n - 1]];
    }
};
